### PR TITLE
Deduplicate the names for the additional domain names

### LINKED ISSUE #3756: duplicated hosts error after configuring the additional domains
/kind bug

**What steps did you take and what happened:**
It is possible that the host is already configured on virtual service for some of the inference services in the cluster as Knative allows exposing domains with labels. We need to add the check before adding the host to the list on virtual service.

https://github.com/kserve/kserve/blob/56a4e7d3c8e4f0a5f9def43a94687c81f6645346/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go#L310

{"level":"error","ts":"2024-06-23T11:22:07Z","msg":"Reconciler error","controller":"inferenceservice","controllerGroup":"serving.kserve.io","controllerKind":"InferenceService","InferenceService":{"name":"inferno-triton","namespace":"s-dsplatform"},"namespace":"s-dsplatform","name":"inferno-triton","reconcileID":"535a7fdd-9af2-42cc-a646-acd30d265dc6","error":"fails to reconcile ingress: fails to create or update ingress: admission webhook \"validation.istio.io\" denied the request: configuration is invalid: duplicate hosts in virtual service: inferno-triton-s-dsplatform.xxxxx & inferno-triton-s-dsplatform.xxxxx","errorVerbose":"admission webhook \"validation.istio.io\" denied the request: configuration is invalid: duplicate hosts in virtual service: inferno-triton-s-dsplatform.xxxxx & inferno-triton-s-dsplatform.xxxxx fails to create or update ingress\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress.(*IngressReconciler).Reconcile\n\t/go/src/github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:590\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice.(*InferenceServiceReconciler).Reconcile\n\t/go/src/github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/controller.go:252\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650\nfails to reconcile ingress\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice.(*InferenceServiceReconciler).Reconcile\n\t/go/src/github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}

**What did you expect to happen:**

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version:
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by houshengbo**
**What this PR does / why we need it**:
- So far, if there are duplicate domain names defined in the additional domain names, or the same domain name defined in both additional domain names and the domain name, it leads to the error `configuration is invalid: duplicate hosts in virtual service`. This PR removes the duplicated domain name for both additional domain names and domain names.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3756

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Comment by greenmoon55** [comment]
Could you fill in the all the sections in the PR description? like `What this PR does / why we need it:`, `Feature/Issue validation/testing` and `Release Note`?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 1]
```diff
@@ -125,14 +125,21 @@ func getAdditionalHosts(domainList *[]string, serviceHost string, config *v1beta
...
+		// do not have duplicate domains.
+		deduplicateMap := map[string]bool{}
 		for _, domain := range *config.AdditionalIngressDomains {
-			host := fmt.Sprintf("%s%s", subdomain, domain)
-			if err := validation.IsDNS1123Subdomain(host); len(err) > 0 {
-				log.Error(fmt.Errorf("The domain name %s in the additionalIngressDomains is not valid", domain),
-					"Failed to get the valid host name")
-				continue
+			// If the domain is redundant, go to the next element.
+			if !deduplicateMap[domain] {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:133]: just to confirm for boolean map this is good enough, we do not need to check the existence flag.

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 2]
```diff
@@ -307,22 +314,49 @@ func createHTTPMatchRequest(prefix, targetHost, internalHost string, additionalH
...
 	return matchRequests
 }
 
+func containsHTTPMatchRequest(matchRequest *istiov1beta1.HTTPMatchRequest, matchRequests []*istiov1beta1.HTTPMatchRequest) bool {
+	for _, matchRequestEle := range matchRequests {
+		// If authority, gateways and uri are all equal, two HTTPMatchRequests will be equal.
+		if stringMatchEqual(matchRequest.Authority, matchRequestEle.Authority) && gatewaysEqual(matchRequest, matchRequestEle) &&
+			stringMatchEqual(matchRequest.Uri, matchRequestEle.Uri) {
+			return true
+		}
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:341]: So we can't perform deepEqual directly on `HTTPMatchRequest` ?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go (hunk 1 above)
Comment (houshengbo) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:133]: Yes, this is good use of the boolean map.
If domain exists, the value is true; if it does not exist, the value is false.

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go (hunk 2 above)
Comment (houshengbo) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:341]: No, I tried. It seems I have to break them and check the field one  by one with this approach.

**Pull request merged**